### PR TITLE
Learning path(fast): Add adaptive serving guide and codelab. Fixes #765, Fixes #665.

### LINKED ISSUE #665: Fast content - Adaptive network serving
This will cover adaptive network serving with with navigator.connection.effectiveType.

### PR DESCRIPTION
**Pull request by robdodson**
Fixes #765, #665.

Changes proposed in this pull request:

- Add adaptive serving guide.
- Add adaptive serving codelab.

Staged at: https://20190501t152948-dot-web-dev-staging.appspot.com/adaptive-serving-based-on-network-quality/

### CODE REVIEW

**Comment by robdodson** [comment]
A few things worth noting:
- I removed smart quotes and replaced them with single quotes. I've had folks accidentally copy/pate smart quotes during a codelab once and they couldn't figure out why their JS wouldn't compile because their editor styled them similarly.
- The link for codelabs appears at the bottom now. You can also add an inline callout for someone to go checkout your codelab by doing this in your markdown:
```
{% Aside 'codelab' %}
Check out [this awesome codelab](link/to/codelab)
{% endAside %}
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+serve an image. `effectiveType` can be `'slow-2g'`, `'2g'`, `'3g'`, or `'4g'`.
+
+## Step 1: Check connection type
+
+The background video is currently specified on line 22 of `index.html` and then
+lazy-loaded in `script.js`.
+
+To load it conditionally, first check if connection type detection is supported.
+In `script.js` add an if statement that tests whether the `navigator.connection`
+object exists and whether it has the `effectiveType` property.
```
[the remark this answers is not in the document]
Comment (arthurevans) [reply src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]: (Recommend putting `if` and `else` in code font.)

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+if (navigator.connection && !!navigator.connection.effectiveType) {
+  if (navigator.connection.effectiveType === '4g') {
+    // video loading code
+  } else {
+    // image loading code
+  }
+}
+```
+
+Wrap the existing video loading code in an else statement, so that video will
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
Wrap the existing video loading code in an `else` statement, so that video will
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+  const videoSource = video.getAttribute('data-src');
+  video.setAttribute('src', videoSource);
+
+  video.setAttribute('style', 'height: 100%; width: 100%; display:inline');
+}
+```
+
+## Step 2: Load video
+
+If the `effectiveType` is `'4g'`, you can use the video loading code from the
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
If the `effectiveType` is `'4g'`, use the video loading code from the
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+The
+[Network Information API](https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation)
+enables you to access information about the user's connection quality. You will
+use its `effectiveType` property to decide when to serve a video and when to
+serve an image. `effectiveType` can be `'slow-2g'`, `'2g'`, `'3g'`, or `'4g'`.
+
+## Step 1: Check connection type
+
+The background video is currently specified on line 22 of `index.html` and then
```
[the remark this answers is not in the document]
Comment (arthurevans) [reply src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]: Also: since there's a longer explanation of this later on, maybe have a reference here, like, "The video loading code is described in more detail in [section]"

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+  const video = document.getElementById('coverVideo');
+  const videoSource = video.getAttribute('data-src');
+  video.setAttribute('src', videoSource);
+  video.setAttribute('style', 'height: 100%; width: 100%; display:inline');
+} else {
+  // image loading code
+}
+```
+
+Here’s how the video loading code works: the video source is specified inside
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
Here’s how the video loading code works: the video source is specified using the
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+[data-src attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)
+in `index.html` and then lazy-loaded in `script.js`.
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
on the `<video>` tag.
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+  const videoSource = video.getAttribute('data-src');
+  video.setAttribute('src', videoSource);
+  video.setAttribute('style', 'height: 100%; width: 100%; display:inline');
+} else {
+  // image loading code
+}
+```
+
+Here’s how the video loading code works: the video source is specified inside
+[data-src attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]: Don't need two links to this doc here. Recommend maybe cutting this one and only including the following one. Also: `data-src` should be in code font.

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md (hunk 5 above)
Comment (arthurevans) [reply src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]: Alternately, maybe explain a little more here. Like:

.. code works: the `<video>` tag doesn't download or display anything at first because its `src` attribute is not set. The video URL to load is specified using the `data-src` attribute.

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+in `index.html` and then lazy-loaded in `script.js`.
+
+```html
+<video id="coverVideo" autoplay loop muted data-src="https://cdn.glitch.com/b6491350-b058-4eb6-aa6c-55c93122073e%2FMatrix%2C%20Console%2C%20Hacking%2C%20Code.mp4?1551464245607"></video>
+```
+
+Because the video source is specified in the `data-src` attribute, it isn’t
+displayed or downloaded initially.
+[Data attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)
+allow you to store extra information on standard HTML elements. They can be
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
allow you to store extra information on standard HTML elements. A data element can be
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+named anything, as long as it starts with "data-".
+
+To actually display the video on the page, the location from `data-src` needs to
+be set as the `src` attribute of the video element. This is done in `script.js`.
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]: "the location from ..." maybe rephrase as "you need to copy the value from `data-src` to the video element's `src` attribute."

I think . you can leave off the "This is done in script.js."

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+  image.setAttribute('style', 'height: 100%; width: 100%; display:inline');
+}
+```
+
+## Try it out
+
+To test it yourself:
+
+1. Mouse over the editor, press the **App** button, then press the **Show**
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]: "press the **App** button  ... " can be shortened to "click **App**, then click **Show** to preview the app."

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+   button to preview the app.
+2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.
+3. Click on the **Network tab**.
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
3. Click the **Network** tab.
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+4. Select **Fast 3G** from throttling presets.
+
+<img class="w-screenshot" src="./devtools_network_throttling.png" alt='DevTools Network tab with Fast 3G throttling option highlighted'>
+
+Now reload the page with Fast 3G still enabled and the app will load an image in the background instead of the video:
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
Now reload the page with Fast 3G still enabled. The app loads an image in the background instead of the video:
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+<img class="w-screenshot" src="./netinfo_app_image.png" alt='Matrix-like image background with "NETWORK INFORMATION" text overlay'>
+
+## Extra Credit: Respond to changes
+
+Remember how this API has an `onchange`
+[event listener](http://localhost:8080/fast/adaptive-serving-based-on-network-quality#how-it-works)?
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]: Bad URL. (I think just cut http://localhost:8080 and make it an absolute path?)

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+You can use it for many things: dynamically adapt content such as video quality, restart deferred data transfers when a change to a high-bandwidth network type is detected, or notify users when the network quality changes.
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
You can use it for many things: dynamically adapting content such as video quality; restarting deferred data transfers when a change to a high-bandwidth network type is detected; or notifying users when the network quality changes.
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+
+Here’s a simple example of how this listener works, add it to `script.js`. This
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
Here’s a simple example of how to use this listener. Add it to `script.js`. This
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+displayNetworkInfo();
+```
+
+Here’s the final state of the [app on Glitch](https://glitch.com/~adaptive-serving-netinfo).
+
+<img class="w-screenshot" src="./netinfo_app_video_background.png" alt='Matrix-like video background with "NETWORK INFORMATION 4g" text overlay'>
+
+To test it again:
+
+1. Mouse over the editor, press the **App** button, then press the **Show**
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
1. Mouse over the editor, click **App** button, then click **Show**
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+   button to preview the app.
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
   to preview the app.
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.
+3. Click on the **Network tab**.
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
3. Click the **Network** tab.
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+
+## Try it out
+
+To test it yourself:
+
+1. Mouse over the editor, press the **App** button, then press the **Show**
+   button to preview the app.
+2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.
+3. Click on the **Network tab**.
+4. Select **Fast 3G** from throttling presets.
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
4. Click the **Throttling** dropdown, which is set to **No throttling** by default. Select  **Fast 3G**.
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+
+<img class="w-screenshot" src="./netinfo_app_video_background.png" alt='Matrix-like video background with "NETWORK INFORMATION 4g" text overlay'>
+
+To test it again:
+
+1. Mouse over the editor, press the **App** button, then press the **Show**
+   button to preview the app.
+2. Press `CTRL + SHIFT + J` or `CMD + OPTION + J` (Mac), to open DevTools.
+3. Click on the **Network tab**.
+4. Select **Fast 3G** from throttling presets.
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
4. Click the **Throttling** dropdown, which is set to **No throttling** by default. Select  **Fast 3G**.
```

File: src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+5. **Reload** the page.
```
Comment (arthurevans) [inline src/site/content/en/fast/codelab-adapt-video-to-image-serving-based-on-network-quality/index.md]:
```suggestion
5. Reload the page.
```

File: src/site/content/en/fast/adaptive-serving-based-on-network-quality/index.md [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+There are many ways you can use this network information to improve the user
+experience:
+
+* Switch between serving high-definition and low-definition content based on the
+  user's network. 
+* Decide whether to preload resources. 
+* Defer uploads and downloads when users are on a slow connection. 
+* Enable offline mode if the network quality is not good enough to load the app
+  and use the features.
+* Warn users that doing something (e.g. watching video) over cellular could cost
```
Comment (arthurevans) [inline src/site/content/en/fast/adaptive-serving-based-on-network-quality/index.md]:
```suggestion
* Warn users that doing something (for example, watching video) over cellular could cost
```

File: src/site/content/en/fast/adaptive-serving-based-on-network-quality/index.md [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+  Use Network Information API to adapt the assets served to users based on the
+  quality of their connection.
+codelabs:
+  - codelab-adapt-video-to-image-serving-based-on-network-quality
+---
+
+Loading a website can be a very different experience depending on the network
+conditions. Everything is usually smooth when you are on a fast network, but
+when you're on the go with a humble data plan and spotty connection, or stuck
+with a laptop on a sluggish coffee shop wifi, it's a different story.
```
[the remark this answers is not in the document]
Comment (robdodson) [reply src/site/content/en/fast/adaptive-serving-based-on-network-quality/index.md]: Added `Wi-Fi` to our guide here: https://github.com/GoogleChrome/web.dev/wiki/Writing-style#spelling-and-capitalisation

**Comment by mihajlija** [comment]
@arthurevans Thank you for the thoughtful review! 😊 I've fixed the formatting and updated the text, you can take another look when you get a chance. 

I don't have the access privileges to build the site locally, so please double-check the links in the codelab which reference the guide (especially the one with the colon, on line 25).

**Comment by robdodson** [comment]
Hey Milica, sorry about that. I just added you to the private repo that
will let you pull in the styles.
https://github.com/GoogleChrome/web.dev-frontend/invitations

*From: *Milica <notifications@github.com>
*Date: *Sun, May 5, 2019 at 4:21 PM
*To: *GoogleChrome/web.dev
*Cc: *Rob Dodson, Author

@arthurevans <https://github.com/arthurevans> Thank you for the thoughtful

**Comment by mihajlija** [comment]
@robdodson Thanks! I've checked the preview, the links in the content work, but returning from a codelab to the associated guide seems to be broken. Looks like the return button points to the old URL scheme. Please ignore this if it's just a preview issue :)

Screen Shot 2019-05-06 at 11 38 55 AM

Screen Shot 2019-05-06 at 11 44 22 AM

**Comment by robdodson** [comment]
@mihajlija ah I think the return button is fixed in current master.

If you checkout the master branch, do a git pull, then checkout your feature branch and run `git rebase master` it should fix it.

Taking a quick look at the post now to see if we can land it.

**Comment by robdodson** [comment]
Looks good to me, let's ship it! 🚢
